### PR TITLE
add substack

### PR DESCRIPTION
**Pull request by 0xvgd**
- Add `EMBED_PLATFORM=substack` to the client env variable

It decides which blog system (notion vs substack) it will render.

- Proxy setting has been added to webpack dev mode

This is because Substack makes api call to the same domain. For example, if you visit https://polymorpher.substack.com, an api call is made to https://polymorpher.substack.com/api/v1/archive. The api call is made automatically from Substack Javascript regardless of our frontend code, which the frontend can't do anything to prevent it.

Hence, we need a way to proxy the request to the actual backend which handles the api. In production mode, we need to configure web server like Nginx or Apache to proxy request `https://app-domain/api/v1/archive`  to `https://app-server/[substack-domain]/api/v1/archive` before the React client handles it.

### CODE REVIEW

File: client/config.ts [hunk 1]
```diff
@@ -3,6 +3,7 @@ const debug = process.env.DEBUG
 const config = {
   debug,
   embedderContract: process.env.EMBEDDER_CONTRACT ?? (debug ? '0x9BC52FBcCcde8cEADAEde51a25dBeD489b201e53' : '0x476e14D956dca898C33262aecC81407242f8431A'),
+  embedPlatform: process.env.EMBED_PLATFORM ?? 'notion',
```
Comment (polymorpher) [inline client/config.ts:6]: use ewsType value read from contract, instead of configuring as fixed value at client

File: client/src/Page.tsx [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+  return <TweetEmbed tweetId={id} />
+}
+
+const sld = getSld()
+const subdomain = getSubdomain()
+const pageIdOverride = parsePath(getPath().slice(1))
+
+const notion = config.embedPlatform === 'notion'
+
+const Page: React.FC = () => {
```
Comment (polymorpher) [inline client/src/Page.tsx]: use separate components for notion and substack - they have very different rendering logic

File: server/routes/index.ts [hunk 3]
```diff
@@ -20,6 +23,21 @@ router.get('/health', async (req, res) => {
...
 })
 
+router.get('/page',
+  limiter(),
+  async (req, res) => {
+    try {
+      if (!req.query.url) {
+        throw new Error('URL query param is not specified')
+      }
+      const { data } = await axiosBase.get(req.query.url as string)
```
Comment (polymorpher) [inline server/routes/index.ts:88]: provide substack specific api, not generalized proxy - we never intended to do that and we need to prevent misuse and abuse

File: client/src/hooks/useTryCatch.tsx [hunk 4]
```diff
@@ -11,7 +11,7 @@ interface TryCatchState {
 }
 export const useTryCatch = (): TryCatchState => {
   const [pending, setPending] = useState(false)
-  const [initializing, setInitializing] = useState(true)
+  const [initializing, setInitializing] = useState(false)
```
Comment (polymorpher) [inline client/src/hooks/useTryCatch.tsx]: need to keep as true otherwise some elements are interactive from the start

**Comment by polymorpher** [comment]
notion is not a blog system, it is used extensively in production for website hosting for .country domains. should not mix it and substack - they serve completely different product purposes

**Comment by polymorpher** [comment]
need to revert all changes made to and relevant to notion.tsx. Implement substack rendering stuff completely separately. Manage page can be shared, but UI need to show clear distinction

File: client/webpack.config.ts [hunk 5]
```diff
@@ -18,6 +18,13 @@ module.exports = {
     client: {
       overlay: false,
       progress: true
+    },
+    // need to be configured via web server like nginx or apache in production mode
+    proxy: {
+      '/api/v1': {
+        target: 'http://localhost:3100',
+        router: () => 'http://localhost:3002/polymorpher.substack.com'
```
Comment (polymorpher) [inline client/webpack.config.ts:26]: shouldn't just point to a fixed substack site since we need to test many configurations in development mode. You can use router and parse the req, but I think you could just use rewrite - it could be a lot simpler

**Comment by polymorpher** [comment]
In production, we can use load balancer path matcher and path rewriting to route api calls under the same domain to an appropriate backend proxy api (at any path of host we want)

File: server/routes/index.ts (hunk 3 above)
Comment (0xvgd) [reply server/routes/index.ts:88]: Okay

File: client/config.ts (hunk 1 above)
Comment (0xvgd) [reply client/config.ts:6]: @polymorpher yes, we can get ewsType for registered domain.
My question is how we can get ewsType in /manage page before registration?
Also we need to tell the smart contract that the domain is for whether notion or substack when calling `client.update`

File: client/src/Page.tsx (hunk 2 above)
Comment (0xvgd) [reply client/src/Page.tsx]: Okay

File: client/src/hooks/useTryCatch.tsx (hunk 4 above)
Comment (0xvgd) [reply client/src/hooks/useTryCatch.tsx]: Okay

File: client/webpack.config.ts (hunk 5 above)
Comment (0xvgd) [reply client/webpack.config.ts:26]: Request `/api/v1` is made automatically from substack script to the default domain the script is running from. For example, on `https://polymorpher.substack.com`, request to `/api/v1` (hence https://polymorpher.substack.com/api/v1) is made. On the other hand, when a substack page is rendered in our client app `sample.com`, the api is made to `sample.com/api/v1` .

And we can't control anything about the request `/api/v1` because it is called from substack script, which means we can't add any parameter specifying which substack domain it is for to the request url from the client app.

Hence, in order to get the substack domain this request is about, I think we need to
- parse the request to get request origin
- query the smart contract with the request origin to get substack domain

from the backend code. Is it okay to put smart contract interaction on the backend side?
It feels pretty complicated code than testing it via webpack proxy, but want your confirmation before doing it.

File: server/routes/index.ts (hunk 3 above)
Comment (0xvgd) [reply server/routes/index.ts:88]: Done

File: client/src/hooks/useTryCatch.tsx (hunk 4 above)
Comment (0xvgd) [reply client/src/hooks/useTryCatch.tsx]: Done

File: client/src/Page.tsx (hunk 2 above)
Comment (0xvgd) [reply client/src/Page.tsx]: Done

File: client/src/AppRoutes.tsx [hunk 6]
```diff
@@ -1,14 +1,18 @@
...
+
+const notion = config.embedPlatform === 'notion'
 
 const AppRoutes: React.FC = () => {
   return (
     <BrowserRouter>
       <Routes>
         <Route path={'/manage'} element={ <Manage />} />
-        <Route path='/*' element={ <Notion />} />
+        <Route path='/*' element={ notion ? <Notion /> : <Substack /> } />
```
Comment (polymorpher) [inline client/src/AppRoutes.tsx:15]: It needs to read from contract to figure out what to render, not from a fixed config

File: client/config.ts (hunk 1 above)
Comment (polymorpher) [reply client/config.ts:6]: Manage page is for setting ewsType of a subdomain config, not getting it. client.update already has the interface for it `update: (sld: string, subdomain: string, ewsType: EWSType, page: string, pages: string[], landingPageOnly: boolean) => Promise<ContractTransaction>`

File: client/src/Manage.tsx [hunk 7]
```diff
@@ -71,12 +72,30 @@ const SuggestedPageId = ({ id, applyId }: SuggestedPageIdConfig): JSX.Element =>
     return <></>
   }
   if (id === null) {
-    return <SmallTextRed>Failed to extract notion page id. Please check the url.</SmallTextRed>
+    return (
+      <SmallTextRed>
+        Failed to extract {config.embedPlatform} page id. Please check the url.
```
Comment (polymorpher) [inline client/src/Manage.tsx]: config.embedPlatform should be removed

File: client/src/Manage.tsx [hunk 8]
```diff
@@ -71,12 +72,30 @@ const SuggestedPageId = ({ id, applyId }: SuggestedPageIdConfig): JSX.Element =>
...
+      </SmallTextRed>
+    )
   }
   if (Object.prototype.toString.call(id) === '[object Error]') {
-    return <SmallTextRed>Unable to parse the url to extract notion page id. Error: {id.toString()} </SmallTextRed>
+    return (
+      <SmallTextRed>
+        Unable to parse the url to extract {config.embedPlatform} page id. Error: {id.toString()}
```
Comment (polymorpher) [inline client/src/Manage.tsx]: same

File: client/src/Manage.tsx [hunk 9]
```diff
@@ -71,12 +72,30 @@ const SuggestedPageId = ({ id, applyId }: SuggestedPageIdConfig): JSX.Element =>
...
+      </SmallTextRed>
+    )
   }
-  return <SmallTextGrey>Extracted notion page id from url: <span style={{ color: 'black', cursor: 'pointer', textDecoration: 'underline' }} onClick={() => { applyId?.(id) }}>{id.toString()}</span></SmallTextGrey>
+  return (
+    <SmallTextGrey>
+      Extracted {config.embedPlatform} page id from url:
```
Comment (polymorpher) [inline client/src/Manage.tsx]: same

File: client/src/Manage.tsx [hunk 10]
```diff
@@ -126,31 +145,77 @@ const Manage = (): JSX.Element => {
     if (!debouncedEditingPageId) {
       return
     }
-    if (isValidNotionPageId(debouncedEditingPageId)) {
-      setSuggestedPageId(undefined)
-      return
+
+    if (config.embedPlatform === 'notion') {
```
Comment (polymorpher) [inline client/src/Manage.tsx]: same

File: client/src/Manage.tsx [hunk 11]
```diff
@@ -126,31 +145,77 @@ const Manage = (): JSX.Element => {
...
+        }
+      }
     }
+
     tryCatch(async () => {
-      const tx = await client.update(sld, subdomain, EWSTypes.EWS_NOTION, pageId, allowedPageIds, false)
+      const tx = await client.update(
+        sld,
+        subdomain,
+        config.embedPlatform === 'notion' ? EWSTypes.EWS_NOTION : EWSTypes.EWS_SUBSTACK,
```
Comment (polymorpher) [inline client/src/Manage.tsx:179]: same

File: client/src/Manage.tsx [hunk 12]
```diff
@@ -126,31 +145,77 @@ const Manage = (): JSX.Element => {
...
 
   const save = async (): Promise<void> => {
-    if (!isValidNotionPageId(pageId) && pageId !== '') {
-      toast.error(`Invalid landing page id: ${pageId}`)
-      return
-    }
-    for (const id of allowedPageIds) {
-      if (!isValidNotionPageId(id)) {
-        toast.error(`Invalid additional page id: ${id}`)
+    if (config.embedPlatform === 'notion') {
```
Comment (polymorpher) [inline client/src/Manage.tsx]: same

File: common/substack-utils.ts [hunk 13]
```diff
@@ -0,0 +1,6 @@
+export const isValidUrlPath = /^\/[/.a-zA-Z0-9-_#]+$/.test
+
+export const getLandingPagePath = (path: string) => {
+  const url = new URL(path)
+  return url.pathname === '/' ? url.origin : null
```
Comment (polymorpher) [inline common/substack-utils.ts:5]: why null?

File: server/src/notion.ts [hunk 14]
```diff
@@ -13,10 +12,9 @@ import {
   extractEmoji,
   extractPageImagePreview
 } from '../../common/notion-utils.ts'
+import { axiosBase } from '../routes/index.ts'
```
Comment (polymorpher) [inline server/src/notion.ts:15]: should not import from route. It will cause circular dependency

File: server/routes/index.ts [hunk 15]
```diff
@@ -129,6 +148,14 @@ router.get('/links',
     }
   })
 
+router.get('/:substackHost/api/v1/archive',
+  limiter(),
+  async (req, res) => {
+    const { data } = await axiosBase.get(`https://${req.params.substackHost}/api/v1/archive`, { params: req.query })
```
Comment (polymorpher) [inline server/routes/index.ts:154]: It's a security risk. Other people can use this API as a proxy to obtain response from arbitrary host at 
path `/api/v1/archive`. This should be restricted to substack subdomain only. If a custom domain is provided, the domain should be first resolved to the equivalent substack subdomain by examining CNAME record (or more direct ways without using DNS)

File: yarn.lock [hunk 16]
```diff
@@ -0,0 +1,4 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (polymorpher) [inline yarn.lock:1]: should be removed

File: server/routes/index.ts (hunk 3 above)
Comment (polymorpher) [reply server/routes/index.ts:88]: still here

File: client/src/Manage.tsx [hunk 17]
```diff
@@ -222,7 +287,7 @@ const Manage = (): JSX.Element => {
           {(editingPageIdPosition === 0) && <SuggestedPageId id={suggestedPageId} applyId={setPageId}/>}
           <SmallTextGrey>This is the landing page when people visit {subdomain}{subdomain ? '.' : ''}{sld}.{config.tld} </SmallTextGrey>
           <LabelText style={{ marginTop: 32 }}>Additional pages</LabelText>
-          <SmallTextGrey>You allow the following subpages on web.{sld}.{config.tld}. Links to these pages will be rewritten under web.{sld}.{config.tld}, instead of to external sites (e.g. https://notion.so/....)</SmallTextGrey>
+          <SmallTextGrey>You allow the following subpages on web.{sld}.{config.tld}. Links to these pages will be rewritten under web.{sld}.{config.tld}, instead of to external sites (e.g. {config.embedPlatform === 'notion' ? 'https://notion.so/....' : 'https://polymorpher.substack.com/...'})</SmallTextGrey>
```
Comment (polymorpher) [inline client/src/Manage.tsx]: same

File: client/src/Substack.tsx [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+const pageIdOverride = parsePath(getPath().slice(1))
+
+const client = buildClient()
+
+const Substack: React.FC = () => {
+  const [page, setPage] = useState<string>()
+  // set "https://polymorpher.substack.com" to `pageId` for test purpose
+  const [pageId, setPageId] = useState<string>()
+  // set ["archive"] to `allowedPageIds` for test purpose
+  const [allowedPageIds, setAllowedPageIds] = useState<string[]>([])
```
Comment (polymorpher) [inline client/src/Substack.tsx]: How is allowed page used in Substack?

File: client/src/Substack.tsx [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+import { getPath, getSld, getSubdomain } from './utils'
+import { useTryCatch } from './hooks/useTryCatch'
+import { Navigate } from 'react-router-dom'
+import { BlankPage, LoadingScreen } from './components/Misc'
+import { LinkWrarpper } from './components/Controls'
+import { FlexColumn } from './components/Layout'
+
+const sld = getSld()
+const subdomain = getSubdomain()
+const pageIdOverride = parsePath(getPath().slice(1))
```
Comment (polymorpher) [inline client/src/Substack.tsx]: may update without page or component reload, should not live outside

File: client/src/Substack.tsx [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+      newDiv.innerHTML = page
+
+      const scripts = Array.from(newDiv.querySelectorAll('script'))
+      const oldScripts = Array.from(window.document.querySelectorAll('script[created-from=substack]'))
+
+      for (const script of oldScripts) {
+        script.remove()
+      }
+
+      for (const script of scripts) {
```
Comment (polymorpher) [inline client/src/Substack.tsx]: why do we remove scripts then add exactly the same scripts here?

File: client/src/Substack.tsx [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+
+      setPage(page)
+    })
+  }, [pageId, allowedPageIds, tryCatch])
+
+  useEffect(() => {
+    return () => {
+      const oldScripts = Array.from(window.document.querySelectorAll('script[created-from=substack]'))
+
+      for (const script of oldScripts) {
```
Comment (polymorpher) [inline client/src/Substack.tsx]: why is this necessary? i.e. remove scripts at page unload

File: client/src/Substack.tsx [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+          <BaseText>
+            This site has not connected with any substack page<br/><br/>
+            If you are the owner, please visit <LinkWrarpper href={'/manage'}>here</LinkWrarpper> to configure the site
+          </BaseText>
+        </FlexColumn>
+      </BlankPage>
+    )
+  }
+
+  if (pageIdOverride && !allowedPageIds.includes(pageIdOverride) && pageIdOverride !== pageId) {
```
Comment (polymorpher) [inline client/src/Substack.tsx]: what would be the semantic meaning of allowedPageId restriction here? Control which blog to display on the same-site? Seems not very useful. What else can we do with allowedPageIds in substack context?

File: client/src/Substack.tsx [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+  if (pending) {
+    return <LoadingScreen/>
+  }
+
+  if (!page) {
+    return <LoadingScreen/>
+  }
+
+  return (
+    <div dangerouslySetInnerHTML={{ __html: page }} />
```
Comment (polymorpher) [inline client/src/Substack.tsx]: we need to make sure the site cannot be used for hosting malicious HTML. Since we are just dumping HTML here from whatever we retrieved at an API (which is acting as a universal proxy), how do we ensure the HTML is safe

File: client/webpack.config.ts (hunk 5 above)
Comment (polymorpher) [reply client/webpack.config.ts:26]: I understood all these. It was self-explanatory and doesn't require elaboration. What I was saying is we can control requests like /api/v1 sent to same origin (where the client is hosted) because we control the load balancer (see my other comment) and can rewrite the request in whatever way we need to, for example, to capture all requests sent to `/api/v1` and route them to a backend (proxy) server instead of the frontend server/bucket, and rewrite the path in URL if we need it. The effect is the same host and different paths are served by different servers. What I was saying above is in development environment (where you don't have a load balancer, unlike production environment), you can use webpack devserver's `pathRewrite`, see https://webpack.js.org/configuration/dev-server/#devserverproxy. This lets you proxy the requests to the local backend server (while preserving everything else, e.g. `host`). For example, set `target: https://localhost:8443` and `pathRewrite: { '^/api': '/coolproxy' }`, then all requests in the browser to`https://hello.localhost:3100/api/v1/blah` (where 3100 is serving the webpack client) will be routed to  `https://localhost:8443/coolproxy/v1/blah` but if you check `req.host` in the route handler of the backend server, you will get `hello.localhost:3100` - so essentially SLD is preserved and you can use it to query smart contrats. I also mentioned you can use router - I was referring to https://github.com/chimurai/http-proxy-middleware#router-objectfunction where you can use very custom logic instead of just a rewrite rule, but I think it is an overkill, like I mentioned before.

**Comment by polymorpher** [comment]
I recommend using a separate Manage component for Substack, because the Substack component seems to be using completely separate logics for processing each user action. It uses different logics to parse url, does not have the concept of page id and strict mode, and does not appear to make use of allowedPages

**Comment by polymorpher** [comment]
Using a separate manage component will also avoid merge conflict, which is a pending issue

**Comment by polymorpher** [comment]
I see, the contract is missing a view function for getting ewsType. Feel free to add one. I will redeploy in production and migrate data later after everything is done

File: client/webpack.config.ts (hunk 5 above)
Comment (0xvgd) [reply client/webpack.config.ts:26]: I see. How about reducing the response time by removing smart contract query and hardcoding substack subdomain when configuring load balancer rewrite rules like
```
/api/v1 => backend-host/substack/polymorpher/api/v1
```
it will query directly to polymorpher.substack.com/api/v1 without querying substack subdomain to the smart contract

File: server/routes/index.ts (hunk 15 above)
Comment (0xvgd) [reply server/routes/index.ts:154]: I see

File: client/src/AppRoutes.tsx (hunk 6 above)
Comment (0xvgd) [reply client/src/AppRoutes.tsx:15]: Yes, we can.
My question is the smart contract query will take some time (of course not much) and since the frontend domain is specific to what to render, we can reduce the load time if we specify what to render directly in the env

File: client/config.ts (hunk 1 above)
Comment (0xvgd) [reply client/config.ts:6]: I see.
My question is how the frontend distinguish what manage page (notion vs substack) it will render before the domain registration. Setting ewsType is done inside Manage component, but don't know ewsType at the time of deciding what manage page to render. I could miss something

Comment (polymorpher) [reply client/config.ts:6]: We could provide a selection button where the user can decide which platform they want the subdomain to connect with (and we can render the corresponding manage component based on user selection). A separate, textbox-based input UI is also implemented, outside this repository.

File: client/webpack.config.ts (hunk 5 above)
Comment (polymorpher) [reply client/webpack.config.ts:26]: We need to have the backend query smart contract based on SLD and subdomain because that's the whole point of doing all these - configure a SLD and its subdomain to connect to a user's substack of choice, not a substack we choose.

File: client/src/AppRoutes.tsx (hunk 6 above)
Comment (polymorpher) [reply client/src/AppRoutes.tsx:15]: All SLDs (i.e. domains) point to the same frontend (client), so you can't decide what to render in the env. It must be determined from the SLD and querying smart contract. See implementation for notion, and test it yourself on some .country domains, e.g. www.h.country, web.zeroknowledge.country, www.thelibrary.country - they all point to the exactly the same frontend deployment (which we deployed, not the user)

File: client/config.ts (hunk 1 above)
Comment (0xvgd) [reply client/config.ts:6]: I see

File: client/webpack.config.ts (hunk 5 above)
Comment (0xvgd) [reply client/webpack.config.ts:26]: Okay

File: client/src/AppRoutes.tsx (hunk 6 above)
Comment (0xvgd) [reply client/src/AppRoutes.tsx:15]: Okay
